hazelcast/src/hazelcast/client/spi/impl/ClientPartitionServiceImpl.cpp

```diff
@@ -0,0 +1,246 @@
+
+                void ClientPartitionServiceImpl::beforeListenerRegister() {
+                }
+
+                void ClientPartitionServiceImpl::onListenerRegister() {
+                }
+
+                boost::shared_ptr<Address> ClientPartitionServiceImpl::getPartitionOwner(int partitionId) {
+                    waitForPartitionsFetchedOnce();
+                    return partitions.get(partitionId);
+                }
+
+                int ClientPartitionServiceImpl::getPartitionId(const serialization::pimpl::Data &key) {
+                    int pc = getPartitionCount();
+                    if (pc <= 0) {
+                        return 0;
+                    }
+                    int hash = key.getPartitionHash();
+                    return util::HashUtil::hashToIndex(hash, pc);
+                }
+
+                int ClientPartitionServiceImpl::getPartitionCount() {
+                    waitForPartitionsFetchedOnce();
+                    return partitionCount;
+                }
+
+                boost::shared_ptr<client::impl::Partition> ClientPartitionServiceImpl::getPartition(int partitionId) {
+                    return boost::shared_ptr<client::impl::Partition>(new PartitionImpl(partitionId, client, *this));
+                }
+
+                void ClientPartitionServiceImpl::waitForPartitionsFetchedOnce() {
+                    while (partitionCount == 0 && client.getConnectionManager().isAlive()) {
+                        if (isClusterFormedByOnlyLiteMembers()) {
+                            throw exception::NoDataMemberInClusterException(
+                                    "ClientPartitionServiceImpl::waitForPartitionsFetchedOnce",
+                                    "Partitions can't be assigned since all nodes in the cluster are lite members");
+                        }
+                        std::auto_ptr<protocol::ClientMessage> requestMessage = protocol::codec::ClientGetPartitionsCodec::RequestParameters::encode();
+                        boost::shared_ptr<ClientInvocation> invocation = ClientInvocation::create(client,
+                                                                                                  requestMessage, "");
+                        boost::shared_ptr<ClientInvocationFuture> future = ClientInvocation::invokeUrgent(invocation);
+                        try {
+                            boost::shared_ptr<protocol::ClientMessage> responseMessage = future->get();
+                            protocol::codec::ClientGetPartitionsCodec::ResponseParameters response =
+                                    protocol::codec::ClientGetPartitionsCodec::ResponseParameters::decode(
+                                            *responseMessage);
+                            // TODO: change with processPartitionResponse(response.partitions, response.partitionStateVersion, response.partitionStateVersionExist)
+                            processPartitionResponse(response.partitions,
```

Comment:
Handling new field should be done, otherwise we will get illegal access when decoding partition service.

Reply:
Done.